core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -245,63 +334,290 @@ public static BytesReference serializeDiffClusterState(Diff diff, Version nodeVe
-                    public void onNewClusterStateFailed(Throwable t) {
-                        try {
-                            channel.sendResponse(t);
-                        } catch (Throwable e) {
-                            logger.debug("failed to send response on cluster state processed", e);
-                        }
+            if (timedout) {
+                markAsFailed("timed out waiting for commit (commit timeout [" + commitTimeout + "]");
+            }
+            if (isCommitted() == false) {
+                throw new Discovery.FailedToCommitClusterStateException("{} enough masters to ack sent cluster state. [{}] left",
+                        timedout ? "timed out while waiting for" : "failed to get", neededMastersToCommit);
+            }
+        }
+
+        synchronized public boolean isCommitted() {
+            return committed;
+        }
+
+        synchronized public void onNodeSendAck(DiscoveryNode node) {
+            if (committed) {
+                assert sendAckedBeforeCommit.isEmpty();
+                sendCommitToNode(node, clusterState, this);
+            } else if (committedOrFailed) {
+                logger.trace("ignoring ack from [{}] for cluster state version [{}]. already failed", node, clusterState.version());
+            } else {
+                // we're still waiting
+                sendAckedBeforeCommit.add(node);
+                if (node.isMasterNode()) {
+                    onMasterNodeSendAck(node);
+                }
+            }
+        }
+
+        synchronized private void onMasterNodeSendAck(DiscoveryNode node) {
+            logger.trace("master node {} acked cluster state version [{}]. processing ... (current pending [{}], needed [{}])",
+                    node, clusterState.version(), pendingMasterNodes, neededMastersToCommit);
+            neededMastersToCommit--;
+            if (neededMastersToCommit == 0) {
+                if (markAsCommitted()) {
+                    for (DiscoveryNode nodeToCommit : sendAckedBeforeCommit) {
+                        sendCommitToNode(nodeToCommit, clusterState, this);
                     }
-                });
-            } catch (Exception e) {
-                logger.warn("unexpected error while processing cluster state version [{}]", e, lastSeenClusterState.version());
-                try {
-                    channel.sendResponse(e);
-                } catch (Throwable e1) {
-                    logger.debug("failed to send response on cluster state processed", e1);
+                    sendAckedBeforeCommit.clear();
                 }
             }
+            onMasterNodeDone(node);
+        }
+
+        synchronized private void onMasterNodeDone(DiscoveryNode node) {
+            pendingMasterNodes--;
+            if (pendingMasterNodes == 0 && neededMastersToCommit > 0) {
+                markAsFailed("no more pending master nodes, but [" + neededMastersToCommit + "] acks are still needed");
```

Comment:
hm. so what I meant in my last comment is that this log message tells what the variables pendingMasterNodes and neededMastersToCommit are set to but not what that means. I think at this point less than a quorum of master eligible nodes have responded with success and the rest failed and therefore we cannot apply the cluster state. Is that right? If so, can we maybe change that to:

`markAsFailed("received responses from all master eligible nodes that are known but less than minimum_master_nodes responded with success and all the others failed ({})", numberOfFailedMasterNodes);` ?